sklearn/datasets/_arff_parser.py

```diff
@@ -69,176 +69,212 @@ def _sparse_data_to_array(
     return y
 
 
-def _feature_to_dtype(feature: Dict[str, str]):
-    """Map feature to dtype for pandas DataFrame"""
-    if feature["data_type"] == "string":
-        return object
-    elif feature["data_type"] == "nominal":
-        return "category"
-    # only numeric, integer, real are left
-    elif feature["number_of_missing_values"] != "0" or feature["data_type"] in [
-        "numeric",
-        "real",
-    ]:
-        # cast to floats when there are any missing values
-        return np.float64
-    elif feature["data_type"] == "integer":
-        return np.int64
-    raise ValueError("Unsupported feature: {}".format(feature))
-
-
-def _convert_arff_data(
-    arff: ArffContainerType,
-    col_slice_x: List[int],
-    col_slice_y: List[int],
-    shape: Optional[Tuple] = None,
-) -> Tuple:
-    """
-    converts the arff object into the appropriate matrix type (np.array or
-    scipy.sparse.csr_matrix) based on the 'data part' (i.e., in the
-    liac-arff dict, the object from the 'data' key)
+def _cast_frame(frame, columns_info):
+    """Cast the columns of a dataframe using the ARFF metadata.
 
     Parameters
     ----------
-    arff : dict
-        As obtained from liac-arff object.
-
-    col_slice_x : list
-        The column indices that are sliced from the original array to return
-        as X data
+    frame : dataframe
+        The dataframe to cast.
 
-    col_slice_y : list
-        The column indices that are sliced from the original array to return
-        as y data
+    columns_info : dict
+        The ARFF metadata for the columns of the dataframe.
 
     Returns
     -------
-    X : np.array or scipy.sparse.csr_matrix
-    y : np.array
+    frame : dataframe
+        The dataframe with the right casting.
     """
-    arff_data = arff["data"]
-    if isinstance(arff_data, Generator):
-        if shape is None:
-            raise ValueError("shape must be provided when arr['data'] is a Generator")
-        if shape[0] == -1:
-            count = -1
+    dtypes = {}
+    for name in frame.columns:
+        column_dtype = columns_info[name]["data_type"]
+        if column_dtype.lower() == "integer":
+            dtypes[name] = "Int64"
```

Comment:
Is an extension array used here because there could be an integer array with missing values?

If we are casting based on the arff metdata, the original parser used `number_of_missing_values` to decide which dtype to use. We can continue to use that here to decide between `Int64` and `np.int64`

Reply:
> Is an extension array used here because there could be an integer array with missing values?

Yes, exactly but we would need to bump the version of pandas.

> the original parser used number_of_missing_values to decide which dtype to use

To be honest, I am concerned with the data returned by the OpenML server. The spec changed over time and depending on if you get an old or a recent dataset then you might not get the same info. It was one of the reasons to rely on such heuristic